text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+
+#### Javascript Files (.js)
+
+No alternatives were considered yet.
+
+#### Type Declaration Files (.d.ts)
+
+Remove these files. We can’t remove the type declaration files, because that
+would break the amazing type-checking features of CDK!
+
+#### Generated L1 Construct files
+
+The auto-generated L1 code overlaps with the Javascript, Source Map, and Type
+Declaration categories already mentioned, and it composes about 25% of the
+overall package. The size of this code grows linearly as the volume of
+CloudFormation resources and features grows. Alternative solutions focusing on
+this category:
+
+1. Separate out the L1 constructs to their own library. This would compromise
+   one of the major values of AWS CDK v2, which is that all AWS CDK constructs
+   are in a single library, and customers no longer have to manage a complicated
+   dependency tree for their CDK apps and libraries. We might consider this in
+   the future as a logical way of breaking up the library into multiple
+   components, if we are not able to keep the size under control with other
+   solutions described in this document.
+2. Invent a new way of generating this code that has a smaller footprint. This
+   could be a huge undertaking, and needs a lot more investigation into whether
+   there are possible solutions for this.
+
+### What are the drawbacks of this solution?
+
+1. This solution does not address the fact that the size of aws-cdk-lib
+   increases with each weekly release, because we are adding code and new
+   features at a rapid pace, and AWS is always adding new services and features.
+   It is possible that after all of the proposed solutions are implemented, that
+   the package size will still continue to increase past a reasonable size. At
+   that point, we might need to invent more creative solutions to reduce the
+   size, or implement some of the alternative solutions outlined here.
+
+2. This solution has a breaking change for some customers. Until we get some
+   feedback from the community on this RFC, we don't know for sure if the
+   proposed migration for those customers is reasonable.
+
+### What is the high-level project plan?
+
+Each category of files can be worked on in parallel and addressed separately.
+The highest priority category to start with will be the Lambda Layer zip files.
+Even though the Lambda Layers are not the biggest contributor to size today,
+they are the biggest potential contributor to size. And, there is customer need
+to add more Lambda Layers, which we are not able to do until the large
+dependencies are not required to be bundled in aws-cdk-lib.
+
+The next category will be the .jsii files. After both of these solutions are
+implemented, the the aws-cdk-lib unpacked npm package size will be about 100 MB.
+
+The lowest priority category will be removing the Source Map files and minifying
+the Type Declaration files. The final size of the package after this work is
+still to be determined.
```

Comment:
It's low priority but let's do it early on anyway, it's an easy win for about 5 minutes of work!

Reply:
Yes, definitely agree with you. This section only describes priority, they can all be done in parallel. So I don't think this plan prevents us from doing it early on. Do I need to specify that?